zowe-rest-api-sample-spring/docs/api-client-authentication.md

```diff
@@ -0,0 +1,201 @@
+This is optional and is meant for diagnostics purposes.
+
+### The support for HTTP basic authentication in the Zowe REST API SDK
+
+REST API services that are developed by the Zowe REST API SDK support HTTP basic authentication by default. When it is running on z/OS then it is using the SAF interface to validate the user ID and password by the z/OS security subsystems. For testing outside of z/OS, it uses a predefined dummy credentials. For more information see [Web Security](web-security.md) and [z/OS Security](zos-security.md).
+
+## Token-based Authentication
+
+Since every REST request requires to provide valid authentication and REST API services do not have mechanism of a long-term web session, it is necessary to have a way how to establish an authenticated session outside of individual REST API services.
+
+In Zowe and in many other applications, this is achieved using JWT tokens that can be obtained by a specialized serviced and then used as the authentication information. This service is described in more details at [Zowe Authentication and Authorization Service](https://github.com/zowe/api-layer/wiki/Zowe-Authentication-and-Authorization-Service).
+
+This section focuses on the way how services in the Zowe API ecosystem are expected to accept and use these tokens so the API clients have unified experience.
+
+### Token-based Login Flow and Request/Response Format
+
+1. The API client obtains JWT token by using POST method on `/auth/login` endpoint of API service that requires valid user ID and password
+
+2. The API service calls an *authentication provider* that returns a JWT token that contains the user ID claim in the HTTP cookie named `apimlAuthenticationToken` with attributes `HttpOnly` and `Secure`.
+
+3. The API client remembers the JWT token or cookie and sends it with every request as the cookie with name `apimlAuthenticationToken`.
+
+#### Obtaining the token
+
+- The full URL is the base URL of the API service plus `/auth/login`. If the application has the base URL with `/api/v1` then the full URL can be: `https://hostname:port/api/v1/auth/login`.
+
+- The credentials are provide in JSON request:
+
+    ```json
+    {
+        "username": "...",
+        "password": "..."
+    }
+    ```
+
+- Successful login returns `204` and empty body with the token in the `apimlAuthenticationToken` cookie
+
+- Failed authentication returns `401` without `WWW-Authenticate`
+
+**Example**:
+
+```bash
+curl -v -c - -X POST "https://localhost:10080/api/v1/auth/login" -d "{ \"username\": \"zowe\", \"password\": \"zowe\"}"
+```
+
+```http
+POST /api/v1/auth/login HTTP/1.1
+Accept: application/json, */*
+Content-Length: 40
+Content-Type: application/json
+
+{
+    "username": "zowe",
+    "password": "zowe"
+}
+
+HTTP/1.1 204
+Set-Cookie: apimlAuthenticationToken=eyJhbGciOiJSUzI1NiJ9...; Path=/; Secure; HttpOnly
+```
+
+#### Authenticated request
+
+The API client just pass the JWT token as a Cookie header with name `apimlAuthenticationToken`:
```

Comment:
Similar to above, why not bearer authorization header?  I have no real opinion, just anticipating the question and want to have an answer ready :-)

Reply:
HTTP-only cookie has advantage for usage in browser. The browser takes care of it automatically and does not allow JavaScript code to access it. 

Authorization Bearer header can be easier to use when another service calls another service.

Actually, Zowe APIML code supports both of them. I wanted to keep just one of them but I think that we should allow both.

Reply:
More details about cookies vs auth headers for JWT - https://www.youtube.com/watch?v=YQzU8xEBiPg&feature=youtu.be&t=1628

Reply:
I have updated the text to allow both of them

Reply:
Thanks for the info